lib/puppet/provider/cisco_route_map/cisco.rb

```diff
@@ -0,0 +1,588 @@
+      next unless @resource[prop]
+      send("#{prop}=", @resource[prop]) if new_rm
+      unless @property_flush[prop].nil?
+        @nu.send("#{prop}=", @property_flush[prop]) if
+          @nu.respond_to?("#{prop}=")
+      end
+    end
+    # custom setters which require one-shot multi-param setters
+    match_community_set
+    match_ext_community_set
+    match_ipv4_multicast_set
+    match_ipv6_multicast_set
+    match_ip_addr_access_list
+    match_ip_addr_prefix_list
+    match_route_type_set
+    set_dampening_set
+    set_distance_set
+    set_community_set
+    set_extcommunity_4bytes_set
+    set_extcommunity_rt_set
+    set_extcommunity_cost_set
+    set_ip_next_hop_set
+    set_ip_precedence
+    set_metric_set
+  end
+
+  def match_ip_addr_access_list
+    pf = @property_flush[:match_ipv4_addr_access_list]
+    v4 = pf.nil? ? @nu.match_ipv4_addr_access_list : pf
+    pf = @property_flush[:match_ipv6_addr_access_list]
+    v6 = pf.nil? ? @nu.match_ipv6_addr_access_list : pf
+    @nu.match_ip_addr_access_list(v4, v6)
+  end
+
+  def match_ip_addr_prefix_list
+    pf = @property_flush[:match_ipv4_addr_prefix_list]
+    v4 = pf.nil? ? @nu.match_ipv4_addr_prefix_list : pf
+    pf = @property_flush[:match_ipv6_addr_prefix_list]
+    v6 = pf.nil? ? @nu.match_ipv6_addr_prefix_list : pf
+    @nu.match_ip_addr_prefix_list(v4, v6)
+  end
+
+  def match_community_set
+    comm = @property_flush[:match_community] ? @property_flush[:match_community] : @nu.match_community
+    pf = @property_flush[:match_community_exact_match]
+    exact = PuppetX::Cisco::Utils.flush_boolean?(pf) ? pf : @nu.match_community_exact_match
+    @nu.match_community_set(comm, exact)
+  end
+
+  def match_ext_community_set
+    comm = @property_flush[:match_ext_community] ? @property_flush[:match_ext_community] : @nu.match_ext_community
+    pf = @property_flush[:match_ext_community_exact_match]
+    exact = PuppetX::Cisco::Utils.flush_boolean?(pf) ? pf : @nu.match_ext_community_exact_match
+    @nu.match_ext_community_set(comm, exact)
+  end
+
+  def match_ipv4_multicast_set
+    attrs = {}
+    vars = [
+      :match_ipv4_multicast_src_addr,
+      :match_ipv4_multicast_group_addr,
+      :match_ipv4_multicast_group_range_begin_addr,
+      :match_ipv4_multicast_group_range_end_addr,
+      :match_ipv4_multicast_rp_addr,
+      :match_ipv4_multicast_rp_type,
+      :match_ipv4_multicast_enable,
+    ]
+    if vars.any? { |p| @property_flush.key?(p) }
+      # At least one var has changed, get all vals from manifest
+      vars.each do |p|
+        if @resource[p] == :default
+          attrs[p] = @nu.send("default_#{p}")
+        else
+          attrs[p] = @resource[p]
+          attrs[p] = PuppetX::Cisco::Utils.bool_sym_to_s(attrs[p])
+        end
+      end
+      @nu.match_ipv4_multicast_set(attrs)
+    else
+      call_empty
+    end
+  end
+
+  # method to keep rubocop happy
+  def call_empty
+  end
+
+  def match_ipv6_multicast_set
```

Comment:
I refactored the `match_set` methods to share common code and remove the requirement on `call_empty`.

```ruby
agent-lab9-pm:beaker_tests:2096> git diff
diff --git a/lib/puppet/provider/cisco_route_map/cisco.rb b/lib/puppet/provider/cisco_route_map/cisco.rb
index 3eab9f1..00d5f33 100644
--- a/lib/puppet/provider/cisco_route_map/cisco.rb
+++ b/lib/puppet/provider/cisco_route_map/cisco.rb
@@ -317,9 +317,24 @@ Puppet::Type.type(:cisco_route_map).provide(:cisco) do
     @nu.match_ext_community_set(comm, exact)
   end
 
-  def match_ipv4_multicast_set
+  def match_set_helper(properties, setter)
+    return unless properties.any? { |p| @property_flush.key?(p) }
     attrs = {}
-    vars = [
+    # At least one var has changed, get all vals from manifest
+    properties.each do |p|
+      if @resource[p] == :default
+        attrs[p] = @nu.send("default_#{p}")
+      else
+        attrs[p] = @resource[p]
+        attrs[p] = PuppetX::Cisco::Utils.bool_sym_to_s(attrs[p])
+      end
+    end
+    @nu.send(setter, *[attrs])
+  end
+
+
+  def match_ipv4_multicast_set
+    properties = [
       :match_ipv4_multicast_src_addr,
       :match_ipv4_multicast_group_addr,
       :match_ipv4_multicast_group_range_begin_addr,
@@ -328,29 +343,11 @@ Puppet::Type.type(:cisco_route_map).provide(:cisco) do
       :match_ipv4_multicast_rp_type,
       :match_ipv4_multicast_enable,
     ]
-    if vars.any? { |p| @property_flush.key?(p) }
-      # At least one var has changed, get all vals from manifest
-      vars.each do |p|
-        if @resource[p] == :default
-          attrs[p] = @nu.send("default_#{p}")
-        else
-          attrs[p] = @resource[p]
-          attrs[p] = PuppetX::Cisco::Utils.bool_sym_to_s(attrs[p])
-        end
-      end
-      @nu.match_ipv4_multicast_set(attrs)
-    else
-      call_empty
-    end
-  end
-
-  # method to keep rubocop happy
-  def call_empty
+    match_set_helper(properties, 'match_ipv4_multicast_set')
   end
 
   def match_ipv6_multicast_set
-    attrs = {}
-    vars = [
+    properties = [
       :match_ipv6_multicast_src_addr,
       :match_ipv6_multicast_group_addr,
       :match_ipv6_multicast_group_range_begin_addr,
@@ -359,25 +356,11 @@ Puppet::Type.type(:cisco_route_map).provide(:cisco) do
       :match_ipv6_multicast_rp_type,
       :match_ipv6_multicast_enable,
     ]
-    if vars.any? { |p| @property_flush.key?(p) }
-      # At least one var has changed, get all vals from manifest
-      vars.each do |p|
-        if @resource[p] == :default
-          attrs[p] = @nu.send("default_#{p}")
-        else
-          attrs[p] = @resource[p]
-          attrs[p] = PuppetX::Cisco::Utils.bool_sym_to_s(attrs[p])
-        end
-      end
-      @nu.match_ipv6_multicast_set(attrs)
-    else
-      call_empty
-    end
+    match_set_helper(properties, 'match_ipv6_multicast_set')
   end
 
   def match_route_type_set
-    attrs = {}
-    vars = [
+    properties = [
       :match_route_type_external,
       :match_route_type_inter_area,
       :match_route_type_internal,
@@ -389,20 +372,7 @@ Puppet::Type.type(:cisco_route_map).provide(:cisco) do
       :match_route_type_type_1,
       :match_route_type_type_2,
     ]
-    if vars.any? { |p| @property_flush.key?(p) }
-      # At least one var has changed, get all vals from manifest
-      vars.each do |p|
-        if @resource[p] == :default
-          attrs[p] = @nu.send("default_#{p}")
-        else
-          attrs[p] = @resource[p]
-          attrs[p] = PuppetX::Cisco::Utils.bool_sym_to_s(attrs[p])
-        end
-      end
-      @nu.match_route_type_set(attrs)
-    else
-      call_empty
-    end
+    match_set_helper(properties, 'match_route_type_set')
   end
 
```

Here are the beaker results on `Everest` with my changes applied.

```puppet
cisco_route_map/test_route_map.rb passed in 478.76 seconds
      Test Suite: tests @ 2017-02-06 13:10:35 -0500

      - Host Configuration Summary -


              - Test Case Summary for suite 'tests' -
       Total Suite Time: 478.76 seconds
      Average Test Time: 478.76 seconds
              Attempted: 1
                 Passed: 1
                 Failed: 0
                Errored: 0
                Skipped: 0
                Pending: 0
                  Total: 1

      - Specific Test Case Status -
        
Failed Tests Cases:
Errored Tests Cases:
Skipped Tests Cases:
Pending Tests Cases:
```